### PR TITLE
Feature/loss gc

### PR DESCRIPTION
**Pull request by mlaurako**
This update addresses issue 23 - Updating Loss and adding Multi-Step Rollout Forecasting (https://github.com/NOAA-EMC/ocelot/issues/23), which implements rollout functionality. Additionally, this code handles temporal boundaries in the data by adjusting the rollout number at the edges.

Here's how it works with 5 training bins and 2 validation bins:

```
bin1: [1,2,3] produces 3 steps with loss calculated on all 3 steps ✓
bin2: [2,3,4] produces 3 steps with loss calculated on all 3 steps ✓
bin3: [3,4,5] produces 3 steps with loss calculated on all 3 steps ✓
bin4: [4,5] produces 2 steps with loss calculated on 2 steps ✓ (at temporal boundary)
bin5: [5] produces 1 step with loss calculated on 1 step ✓ (at temporal boundary)
```

In production mode, the model generates the full requested rollout using preset projection locations. The validation code reflects this production behavior by generating the full rollout using the last time bin's location. However, only predictions with available ground truth are used for validation (loss and other evaluation metrics calculation).

```
bin6: 3-step rollout [6, 7, 7] with 2 ground truth steps [6, 7] results in loss calculated on 2 steps, with NaN placeholders logged for step 3
bin7: 3-step rollout [7, 7, 7] with 1 ground truth step [7] results in loss calculated on 1 step, with NaN placeholders logged for steps 2 and 3
```

Note that logging NaN placeholders is essential to prevent GPU hanging when `sync_dist=True` is enabled for logging.

Planned Enhancements:
1. Fixed global grid for all prediction modes (validation, testing, production)
2. An option for feeding forecast outputs back to the next prediction window (similar to GraphCast, an optional function in GraphDOP)

### CODE REVIEW

**Comment by mlaurako** [comment]
To run the script without rollout (rollout = 1), configure the following settings in train_gnn.py:
```
max_rollout_steps = 1
rollout_schedule = 'fixed' 
```

When using rollout = 1, the performance is comparable to the main branch, and the runtime remains the same (approximately 30 minutes for 1 month of ATMS data).

Screenshot 2025-07-03 at 4 13 17 PM

**Comment by mlaurako** [comment]
image

The results demonstrate that the rollout function is functioning correctly. It's worth noting that GraphCast and GraphDOP followed a specific training approach: they avoided rollout training for the first 300,000 and 63,000 gradient steps respectively, then gradually increased the rollout steps by 1 every 1000 and 1500 gradient steps. While this current implementation proves the core functionality and confirms the model can learn from rollout training, using a fixed number of rollout steps from the start of training isn't ideal. We'll need to conduct additional testing once we address the out-of-memory (OOM) issues.

**Comment by mlaurako** [comment]
OOM limit (7 day data)
```
Deep (hidden = 128, num_layer = 6)
2 rollout, ATMS

Medium (hidden = 96, num_layer = 6)
3 rollout, ATMS

Shallow (hidden = 64, num_layer = 4)
3 rollout, ATMS+ PRES

```

File: gnn_model/gnn_datamodule.py [hunk 1]
```diff
@@ -84,9 +126,14 @@ def _to_data(self, data_dict):
...
-            target_lon_deg=torch.tensor(data_dict["target_lon_deg"], dtype=torch.float32),
-            instrument_ids=data_dict["target_instrument_ids"]
+            instrument_ids=data_dict["target_instrument_ids"],  # MK: should this be named as target_instrument_ids?
+            bin_name=data_dict["bin_name"],
+            target_lat_deg=tensor_conversion(data_dict["target_lat_deg"], dtype=torch.float32),
+            target_lon_deg=tensor_conversion(data_dict["target_lon_deg"], dtype=torch.float32),
+            target_metadata=data_dict["target_metadata"],
+            # target_lat_deg=torch.tensor(data_dict["target_lat_deg"], dtype=torch.float32),
+            # target_lon_deg=torch.tensor(data_dict["target_lon_deg"], dtype=torch.float32),
+            # target_metadata=torch.tensor(data_dict["target_metadata"], dtype=torch.float32),
```
Comment (xincjin-NOAA) [inline gnn_model/gnn_datamodule.py:136]: Do these lines still needed?

File: gnn_model/gnn_datamodule.py [hunk 2]
```diff
@@ -321,10 +374,11 @@ def _create_data_object(self, data_dict):
...
         """
-        if "instrument_ids" in data_dict:
-            instrument_ids = torch.tensor(data_dict["instrument_ids"], dtype=torch.long)
-        else:
-            instrument_ids = None
+        # MK: this seems a redundency?
+        # if "instrument_ids" in data_dict:
+        #     instrument_ids = torch.tensor(data_dict["instrument_ids"], dtype=torch.long)
+        # else:
+        #     instrument_ids = None
```
Comment (xincjin-NOAA) [inline gnn_model/gnn_datamodule.py:381]: You are right just remove them.

File: gnn_model/gnn_model.py [hunk 3]
```diff
@@ -164,236 +192,405 @@ def forward(self, data):
         self.debug(f"[DEBUG] src_encoder max: {src_encoder.max().item()}, min: {src_encoder.min().item()}")
         assert src_encoder.max().item() < x.shape[0], "[ERROR] src_encoder has out-of-bounds indices"
 
-        # === Mask edges whose target mesh nodes are not used downstream
-        used_mesh_nodes = torch.cat([edge_index_decoder[0], edge_index_processor[0], edge_index_processor[1]]).unique()
+        # === Get decoder edges for first step to determine used mesh nodes
+        # edge_index_decoder = data.edge_index_decoder
```
Comment (rmclaren) [inline gnn_model/gnn_model.py:196]: commented code

File: gnn_model/gnn_model.py [hunk 4]
```diff
@@ -164,236 +192,405 @@ def forward(self, data):
...
+            decoder_mesh_nodes = step_data['edge_index_decoder'][0].unique()
+            all_decoder_mesh_nodes.update(decoder_mesh_nodes.cpu().numpy())
 
-        edge_mask = torch.isin(tgt_encoder, used_mesh_nodes)
+        # Get all unique processor nodes
+        processor_mesh_nodes = edge_index_processor.flatten().unique()
+
+        # === Mask edges whose target mesh nodes are not used downstream
+        # put all unique node IDs together:
+        # used_mesh_nodes = torch.cat([edge_index_decoder[0], edge_index_processor[0], edge_index_processor[1]]).unique()
```
Comment (rmclaren) [inline gnn_model/gnn_model.py:208]: ..

File: gnn_model/gnn_model.py [hunk 5]
```diff
@@ -164,236 +192,405 @@ def forward(self, data):
...
+        all_used_mesh_nodes = torch.cat([
+            torch.tensor(list(all_decoder_mesh_nodes), device=data.x.device),
+            processor_mesh_nodes
+        ]).unique()
+
+        # Filter encoder edges to only connect to mesh nodes we'll actually use
+        # edge_mask = torch.isin(tgt_encoder, used_mesh_nodes)
```
Comment (rmclaren) [inline gnn_model/gnn_model.py:215]: ..

**Review by rmclaren — COMMENTED** [review]
Stylistic Comment: There is a very high level of nesting in some of this code, especially gnn_model.py (like the forward function step loop). This makes the code hard to follow as you have to keep track all the levels of if statements try statements and for loops, etc. I might help to employ some de-nesting strategies, such as splitting up the work into multiple  functions instead of one big function... 

Some functions are very long, and could be modules or classes in their own right.

I don't mean to trigger a big refactoring, just some comments I had maybe to keep in mind for the future.

**Review by rmclaren — APPROVED** [review]

File: gnn_model/gnn_model.py [hunk 6]
```diff
@@ -405,34 +602,96 @@ def training_step(self, batch, batch_idx):
...
-            # loss = self.loss_fn(y_pred, y_true)
-            loss = self.huber(y_pred, y_true).mean()  # fallback
+        # Compute loss for all steps
+        total_loss = 0.0
+        for step, (prediction, ground_truth) in enumerate(zip(predictions, ground_truths)):
+            if prediction.shape[0] == ground_truth.shape[0]:
+                # Use step-specific instrument IDs instead of batch.instrument_ids
+                step_instrument_ids = step_data_list[step]['target_instrument_ids']
+
+                if self.use_ocelot_loss:
```
[the remark this answers is not in the document]
Comment (mlaurako) [reply gnn_model/gnn_model.py:663]: Great suggestion! Perhaps we could keep the loss function as is for now and implement this change in a separate issue? What do you think @azadeh-gh ?

**Review by xincjin-NOAA — APPROVED** [review]

**Comment by azadeh-gh** [comment]
@mlaurako Thanks for adding the rollout capability to the code. I’m going to merge the current changes so we can move forward with the next step—adding observation-based encoders and decoders. However, since we’ll need to run the model over longer time periods with more rollout steps, please continue working on improving that part.
